Controllability

I--P opened the pull request:
A simple PR with two functions: ctrb() and obsv() to compute the controllability and the observability matrices of a linear system. The code is rather straightforward I'm more worried about the docstring as I'm not used to explain this in english.
ctrb() could be helpful for PR #4295 also as it would allow to warn the user if he tries to place poles on an (nearly) uncontrollable system.

rgommers commented:
I realize that the names are simply copied from Matlab, but they're so non-descriptive that I'd rather see them spelled out (``observability_matrix``).

http://nl.mathworks.com/help/control/ref/obsv.html says that that function is only there for educational purposes and not for serious control design. So I'm not sure that we should add this. ``ctrb`` also has a disclaimer and says that ``ctrbf`` (staircase algorithm) is better.

@I--P what would you say are the pros/cons for adding this?

I--P commented:
Regarding the names I used those from Python Control Library, I was not aware they were the same in MATLAB.

Anyway, I'm using this only for "Control 101" tutorial classes so Matlab is right regarding the educational purpose, I saw Python Control Library had these functions, I knew how to code them and thought they might be useful at least to raise a warning in my code from PR 4295, end of story. I don't really have any other pros/cons.

rgommers commented:
@WarrenWeckesser @WarrenWeckesser @Eric89GXL opinions on adding these two functions?

Eric89GXL commented:
I'm worried that people might think these functions are suitable for serious control design problems. Perhaps some strongly worded phrasing in the docstring would be enough for now, though. If someone later wants to implement a more rigorous algorithm (staircase?) then we could add a `method='staircase'` kwarg and update the docstring to say that staircase should be used for increased precision. We'd be stuck with having the weaker algorithm as the default for backward compatibility, but that's not the end of the world.

I--P commented:
Or we can wait and I can try to implement staircase if I find some time to do it one of these days.

I--P commented:
I've got a working prototype of the staircase algorithm.

For some reason Matlab's ctrbf outputs the transpose of the results I get, do you think I should:

a/ modify my results to stick to what Matlab users are used to or
b/ output my results as is and let people read the f... fine manual about how Scipy does this ?

Source: https://books.google.fr/books?id=eI2pYHWCVbEC&pg=PA174&lpg=PA174&dq=triangularize+the+matrix+B+using+the+QR+factorization+with+column+pivoting&source=bl&ots=U338ihrVqj&sig=ZstVkOQnVuGxTfbp7hhVYrGQ53s&hl=en&sa=X&ei=PS_rVPLEGsKqU-Prg8gC&ved=0CCIQ6AEwAA#v=onepage&q=triangularize%20the%20matrix%20B%20using%20the%20QR%20factorization%20with%20column%20pivoting&f=false

rgommers commented:
@I--P nice! Regarding comparison to Matlab: if your implementation matches the literature I'd say document it clearly, but if the literature doesn't have a clear opinion about whether to use your form or the transpose then maybe it's better to just be consistent with Matlab.

I--P commented:
I was wrong, Matlab result is not a transpose but I think in the end the results are equivalent (well I do hope so given how Matlab is widespread !).

Anyway I've tested the controllability index given by my implementation of controller_Hessenberg against Octave and I always have the same result on random controllable or non controllable systems.

What bothers me:

- the result on orthogonnal similarity transformation P, therefore on H, Bbar and so on, depends on the way the staircase algorithm is coded, so for the same input Matlab does not have the same output as Octave and I most of the time don't have the same result as any of them (of course I'm taking into account the fact that Matlab's P matrix does not have the same ordering as Octave's or mine). => I have no reliable way to test if my results are OK, I can only test that my implementation outputs the right controllability index. If you have any idea on how to improve this...

- I guess it won't change anything in the end but if this PR is accepted I'd like to have a clear answer to my question in the comments line 1223 of ltisys.py before this is merged. I'm afraid I'm a bit too rusty on linear algebra to fint it by myself.

I--P commented:
More than six months have passed since my last update on this PR. Is there a concensus on whether I should keep coding it or not, i.e is it needed in scipy ?

rgommers commented:
@I--P I think that now that you have the staircase algorithm, this is a good addition to ``scipy.signal``. That's also what I think @Eric89GXL says above (please correct me if I'm wrong Eric). 

I had a quick look at your question in the comment above the ``break`` statement - that'll take me more than the 5 minutes I looked at it to figure out though.

rgommers commented:
Sorry that you didn't get much feedback Irvin; ``scipy.signal`` could do with a bit more maintainer power, and I've been super busy in the second half of last year.

I--P commented:
I'll need time to get back into this too, so take your time regarding my question.

Eric89GXL commented:
@rgommers @I--P yeah I think with the staircase it's a good add.

I--P commented:
Rebased, squashed and hopefully all tests will pass.

Some thoughts:
- short function names are not verbose enough as stated last year but here I think they are way too long. I'm afraid everybody will write "from signal import observability_matrix as obsv"...
- I had no time to do linear algebra, this is only a maintenance commit so my questions from last year are still open
- do these function really belong in a namespace called signal ? FWIW I think that the three function from this PR + place_poles should belong to a namespace called "control"

Eric89GXL commented:
> I'm afraid everybody will write "from signal import observability_matrix as obsv"...

I actually don't mind this particular name, even if people use this import pattern a majority of the time. It makes explicit what is being imported and used.

No opinion on the other points from me.

Eric89GXL commented on scipy/signal/__init__.py:
```diff
@@ -147,6 +147,7 @@
 .. autosummary::
    :toctree: generated/
 
+<<<<<<< HEAD
```
likes like something went wrong during rebase

I--P commented:
OK @Eric89GXL 
Another thought, should controller_Hessenberg return a Bunch as place_poles ?

I--P commented:
After a quick review, the way I compute control_index it completely wrong, please don't merge that.

I--P commented:
It should be OK now.
To elaborate a bit, the numerical result was the right one but the way to compute it was not the best one (to say the least...). I was computing the rank of a submatrix using SVD which could yield some numerical instabilities, the publication advised to use QR with pivoting which is what I did eventually.

Interestingly I now have results similar to octave/slicot() for each and every matrix I return so I guess that's how they do it too, before it was sort of scrambled even it the results were still having the expected properties.

I have some test code which outputs matlab like code fed into octave and all results for tens of thousands of random matrices are the same with an absolute tolerance of 1e-8 and a relative one of 1e-5.

If you have any idea of other tests let me now...

Current test algorithm:
1/ loop several tens of thousand times:
2/ create random A,B,C matrix of reasonnable size

3/ Tests done inside on the scipy side:
```python
    H,Bbar,Cbar,P,l=controller_Hessenberg(A,B,C) 
    assert np.allclose(P.dot(A.dot(P.T)), H)
    assert np.allclose(P.dot(B), Bbar)
    assert np.allclose(np.linalg.inv(P), P.T)
```
4/ dump of H, Bbar, Cbar, P and l into a matlab compatible syntax

Tests done with octave:
```matlab
[H,Bbar,Cbar,T,K]=ctrbf(A,B,C);
assert (py_k==K);
assert (abs(H - H_p) <= (atol + rtol * abs(H_p)))
assert (abs(Bbar - Bbar_p) <= (atol + rtol * abs(Bbar_p)))
assert (abs(Cbar - Cbar_p) <= (atol + rtol * abs(Cbar_p)))
```

With H_p, Bbar_p, C_bar_p,  py_k, the result of the last call to controller_Hessenberg() (i.e H, Bbar, Cbar, P and l)

I--P commented:
I don't understand why the refguide check suddendly fails...

Eric89GXL commented:
@I--P there was probably an upstream update that improved the ref checker, so now it finds problems that were always there but went undetected before.

It looks in particular like there are problems with your formatting of equations -- it thinks they're badly formatted headers. You could look through the code for other examples of such equations to see how they get them to look nice and valid.

Eric89GXL commented:
Also it looks like there's a missing newline somewhere in there.

I--P commented:
Thanks I'll fix that.
Is there a way to check only the files I've modified ? I tried python runtests.py --refguide-check -v -t scipy/signal/tests/test_ltisys.py  with no luck.

Eric89GXL commented:
I don't know much about `runtests.py` but perhaps someone else does. Does it take a while to run for you?

I--P commented:
For refguide-check several minutes as it tests all the codebase, for my own checks only a few seconds as I can specify which tests to run. This is really annoying to say the least...

I--P commented:
This time it will be OK I guess.
Anyway this is my "I-am-clueless day", I don't know how to build the html output of my docstring using sphinx, I've tried **make html** in the *doc/* subfolder but it faits somewhere in the *optimize/* subdirectory.

ev-br commented:
To run the refguide check locally, do `$ python runtests --refguide-check -s signal`.

ev-br commented:
Or run `python setup.py install` and run `tools/refguide_check.py` directly (but this will *install* whatever you have checked out at the moment).

To build the html docs, do `$ cd doc; make html-scipyorg`, but that takes a while.

ev-br commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+def controller_Hessenberg(A,B, C=None, tol=None):
+    r"""
+    Reduce the linear system ``AX + BU`` to the controller Hessenberg form
+    H, B_bar where P*A*P.T = H and P*B = Bbar with P being an orthogonnal
```
I *think* this will not render properly: asterisks have special meaning for Sphinx. You probably want to either wrap it it in double backticks (i.e. make it render as code) or use `:math:`...`` (so that it's TeX math).

ev-br replied:
Also please keep the format of: A single line being the short summary; an empty line; a longer summary. Rationale: the very first line ends up in the summaries, and if it's longer than one line, they get weird.

I--P replied:
Should I keep Foobar in the docstring or switch to ::math:: \bar Foo ?

ev-br replied:
FWIW, I would avoid bars, tildes and hats as much as reasonable, both in code and maths. But it's a matter of taste anyway

ev-br commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+    
+    .. math:: 
+
+        H = \left [ \begin{array}{cc}
```
Not sure how the parser reacts to the space between `\left` and the bracket.

ev-br commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+        
+    tol: float, optional
+        Threshold to detect rank deficiency. See Notes from
+        numpy.linalg.matrix_rank. Default is None which is equivalent to
```
single backticks would turn `numpy.linalg.matrix_rank` into a hyperlink.

I--P replied:
I should have deleted this, this argument is now useless.

ev-br commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+    Parameters
+    ----------
+    A, B : ndarray
+        State-space representation of linear system ``AX + BU``.
```
Format: Please check that `A, B` is understood by numpydoc.

I--P replied:
Iirc it was OK in place_poles.  http://docs.scipy.org/doc/scipy-0.16.1/reference/generated/scipy.signal.place_poles.html

ev-br commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+
+    Cbar : ndarray    
+        Optionnaly if C (output matrix) is given the function will also return
+        Cbar = C*P otherwise Cbar = None will be returned. Cbar has the
```
asterisk

ev-br commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+    uses tb01ud from Slicot and all results were similar using the default
+    tolerances of ``np.allcose()``.
+    
+    Reference
```
Please use Reference<b>s</b>: Section names have special meaning for numpydoc.

ev-br commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+        
+    Notes
+    -----
+    The algorithm comes from [1]. The output from this function has been tested
```
needs a trailing underscore for the reference, [1]_.

ev-br commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+    state-space the system might be very difficult, if not impossible, to 
+    control, its condition number should also be checked. ONE SHOULD ONLY
+    RELY ON THE RANK OF THIS MATRIX FOR EDUCATIONAL PURPOSES DUE TO THE
+    NUMERICAL INSTABILITY OF THIS COMPUTATION.
```
Yuk, all caps. How about using bold or italics font instead if really needs be?

ev-br commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+    control, its condition number should also be checked. ONE SHOULD ONLY
+    RELY ON THE RANK OF THIS MATRIX FOR EDUCATIONAL PURPOSES DUE TO THE
+    NUMERICAL INSTABILITY OF THIS COMPUTATION.
+    For serious control design controller_Hessenberg is the way to go.
```
undescore has a special meaning for Sphinx. If it's a reference to a routine, single backticks are the way to go.

ev-br commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+        State-space representation of linear system ``AX + BU``.
+        
+    C : ndarray, optionnal
+        The output matrix of the linear system (i.e ''Y = CX'').
```
Quotation marks meant to be double backticks?

WarrenWeckesser commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,218 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
     return full_state_feedback
+
+
+def controller_Hessenberg(A,B, C=None, tol=None):
```
Drive-by bike-shedding:  Lower case 'h': `controller_hessenberg`.   (Also, add a space after `A,`.)

Eric89GXL replied:
I have a new favorite term, thanks @WarrenWeckesser. +1 for lower case even though it's a proper name, esp. if that's what's done elsewhere in `scipy`

I--P commented:
I've never read or heard "drive-by bikeshedding", what does it mean ? 
For all the other comments I'll fix that tomorrow.

I--P commented:
Docstring should be prettier now. Here is a link to how it renders:
http://i--p.github.io/html-scipyorg/index.html

Eric89GXL commented:
@I--P the "bike shed" refers to a popular post that described the tendency of people to voice opinions on things that are easy to have an opinion on (and not on things that aren't), see:

https://www.freebsd.org/doc/en_US.ISO8859-1/books/faq/misc.html#idp60701008

So @WarrenWeckesser was making a bit of a joke :)

The rendering looks good to me, and it looks like you've made the CIs happy. Your tests sound reasonable to me. Is this ready for merge from your end, then?

I--P commented:
@Eric89GXL  I understood it was some kind of joke but I didn't get it, now I'll use that :-)
If no one has some other tests to add to what I explained in https://github.com/scipy/scipy/pull/4515#issuecomment-174477210 it is ready for merge imho.

Maybe @rgommers has something to say about whether we should return a bunch or not ?

I--P commented:
As no new feedback seems to be coming I think this is ready for merge.

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,229 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+    The algorithm comes from [1]_. The output from this function has been tested
+    with random controllable and uncontrollable system against Octave which
+    uses tb01ud from Slicot and all results were similar using the default
+    tolerances of `allcose`.
```
I'd move this note about "tested against ...." to the unit tests. We regularly test against other software (R, Matlab, Octave, etc.) and normally don't mention this in a docstring.

rgommers replied:
Also a typo: ``allcose`` --> ``allclose``

rgommers commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2496,229 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+    else:
+        Cbar = np.dot(C,P.T)
+    return H,Bbar,Cbar,P,control_index
+    
```
nitpick: PEP8 would like 2 spaces here

rgommers replied:
and more general PEP8: in many lines there are spaces missing behind commas.

I--P replied:
I must do something wrong here, I always run pep8 before pushing, is there a specific option I should add to detect these issues ?

rgommers replied:
The issue is that there's a lot of old (even pre-PEP8) code that doesn't comply to all PEP8 rules. And we don't want to change all that at once, because (a) it's a lot of work, and (b) it ruins the code history. So there are a number of PEP8 errors excluded from the check in https://github.com/scipy/scipy/blob/master/tox.ini#L41. If you run ``pep8`` locally from the scipy repo root, it'll pick up those excludes. 

It's actually quite difficult to get ``pep8`` to ignore ``tox.ini``, but if I delele it then ``pep8 ltisys.py`` gives a lot of errors. A better alternative is ``flake8``, that doesn't use ``tox.ini`` it looks like. So try ``flake8 ltisys.py`` to see what errors you added (it's not clean on master, but much shorter than on this PR).

rgommers commented on scipy/signal/tests/test_ltisys.py:
```diff
@@ -8,11 +8,62 @@
+        assert_equal(2,c_idx)
+        assert np.allclose(P.dot(A.dot(P.T)), H)
+        assert np.allclose(P.dot(B), Bbar)
+        assert np.allclose(np.linalg.inv(P), P.T)
```
No plain ``assert``s should be used in testcases or code, because those are stripped when byte-compiling with ``-O``.

In this case, use ``assert_allclose`` instead of ``assert np.allclose``.

rgommers commented:
> Maybe @rgommers has something to say about whether we should return a bunch or not ?

A bunch return for ``controller_hessenberg`` would be very good to have. 

For the other two functions there's only a single return value, and as far as I can tell little chance that we'd want to add additional return values later on (correct?). So for those, a bunch wouldn't help.

I--P commented:
Thanks @rgommers, I'll fix this asap.

I--P commented:
@rgommers: pep8 compliance should be better, controller_hessenberg returns a bunch, and I've added an example. Docstring rendering is here: http://i--p.github.io/html-scipyorg/index.html

I--P commented:
I changed the tests, I was never checking the results against known values from other sources.

I--P commented:
Two last questions:
1/ I'm still not sure whether to use SVD or QR with pivoting.
- Book says that SVD is "the best way to do this" but then shows the algorithm with QR (written in bold p176, link above)
- Using QR gives the exact same output as Octave so it's easy to check there is no bug, whereas using SVD gives Abar, Bbar, Cbar, P with the expected properties but we have no reference to test against.
- tb01ud (octave/slicot) uses QR: http://www.icm.tu-bs.de/NICONET/doc/TB01UD.html

2/ I forgot my second question (!)

I--P commented:
The last failures are really weird, the very same test yields the right result on my computer.
Line 2657 of ltisys.py there is:
        if np.allclose(H_sub, 0):  # or H_k_k-1" is a zero matrix
Which is the condition required to stop the algorithm, here it stops with Hsub = -1.13311663e-16 and control_index=2, on travis I get  [[ 0.81649658 -0.28867513]] with control_index = 2 and then an empty Hsub with control_index = 2...

I--P commented:
Got it, I was using R.any() to find a zero line in R after a QR decomposition instead of using isclose().
I'm wondering whether I should add an atol/rtol set of arguments or not... ?

Eric89GXL commented:
Yeah sometimes you need `atol` / `rtol` in `assert_allclose` to get the comparisons right and compensate for floating point arithmetic errors.

Eric89GXL commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2497,271 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+
+    """
+
+    if A.ndim != 2:
```
You should probably precede these with `A = np.asarray(A)` and `B = np.asarray(B)` to ensure you actually have `.ndim` properties available.

Eric89GXL commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2497,271 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+    r"""
+    Compute the observability matrix of a linear system.
+
+    Compute the observability matrix of the linear system:
```
Instead of repeating the same text here, you can say `The observability is defined as:`

Eric89GXL commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2497,271 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+    numerical instability of this computation.**
+    """
+
+    if C.ndim != 2:
```
same thing here about `np.asarray`

Eric89GXL commented on scipy/signal/tests/test_ltisys.py:
```diff
@@ -8,11 +8,123 @@
+        B = np.array([[1, 1], [1, 1], [1, 1]])
+
+        ref_Abar = np.array([[2.3333, -0.4714, 0], [0.9428, 0.6667, 0],
+                             [0, 0, 0]])
```
... and then this can just be `ref_Abar = [[...]]`, i.e. no `np.array()` around it.

Eric89GXL commented on scipy/signal/tests/test_ltisys.py:
```diff
@@ -8,11 +8,123 @@
+        A = np.array([0.7665, 0.1665, 0.9047, 0.4540, 0.5007, 0.4777, 0.4865,
+                      0.5045, 0.2661, 0.3841, 0.2378, 0.8977, 0.5163, 0.0907,
+                      0.2771, 0.2749, 0.9092, 0.3190, 0.9478, 0.9138, 0.3593,
+                      0.0606, 0.9866, 0.0737, 0.5297]).reshape((5, 5))
```
same with these, you won't need the `np.array()` anymore

Eric89GXL replied:
oh, except later you do `A.dot()` and would need to change it to `np.dot(A, ...)` so maybe not worth making the change here

I--P replied:
I like to input big arrays like that, this way I don't have to wonder where to put the [] stuff, all I need is a reshape. Admittedly  it is less readable but I tend to think that if someone really needs to check this array all he has to do is to download the original publication and he will see it in a much more readable way.
However you're right about the array() before trying ndim, I'll fix that.

Eric89GXL commented on scipy/signal/tests/test_ltisys.py:
```diff
@@ -233,7 +345,6 @@ def test_errors(self):
         assert_raises(ValueError, place_poles, A, B,
                       (-2+1j,-2-1j,-2+3j,-2-3j), method="KNV0")
 
-
```
this should probably still be here for PEP8 (two lines before class or function def at root indentation)

I--P replied:
I don't know why/how I managed to remove this line... Thanks for pointing it out.

Eric89GXL commented:
Otherwise LGTM.

I--P commented:
@Eric89GXL the atol/rtol arguments where not for the tests, I was wondering if I should add these in the main function to allow the user to override default atol/rtol from allclose() on lines 2662 and 2632.

Eric89GXL commented on scipy/signal/ltisys.py:
```diff
@@ -2494,3 +2497,279 @@ def place_poles(A, B, poles, method="YT", rtol=1e-3, maxiter=30):
+        # find the first occurence of a zero line in R, this will
+        # reveal the rank of H_sub
+        zero_lines_idx = np.where(np.all(
+                            np.isclose(R, [0]*R.shape[1]), axis=1))[0]
```
I don't think you actually need to do `[0] * R.shape[1]`, doing `np.isclose(R, 0)` should broadcast

Eric89GXL commented:
+1 for adding `tol=1e-8` kwarg. You don't need `rtol` because both comparisons are to a desired value of zero, so just use `np.isclose(..., atol=tol)` and `np.allclose(..., atol=tol)` and you should be good.

I--P commented:
Added atol kwarg, not sure if the test is really relevant though...

I--P commented:
Why not atol ? This way the user knows what kind of tolerance we're using, whereas tol is ambiguous if you don't read the full docstring.

Eric89GXL commented:
One is listed as `tol` and the other as `atol` so they need to at least match. I slightly prefer `tol` because it's the one, single tolerance value that is used -- and when there is only one tolerance, it's generally assumed to be an absolute tolerance level, e.g. first couple results for google search for `numpy tol`:

http://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.rate.html
http://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.linalg.matrix_rank.html

I--P commented:
I didn't notice I mixed atol and tol !
Anyway let's go for tol.

Eric89GXL commented:
@I--P is this good to go now?

I--P commented:
Imho yes.

Eric89GXL commented:
@ev-br @rgommers feel free to have a look if you have time, otherwise I'll take a final look and merge in a couple of days

I--P commented:
I'd really like to have a 2nd, 3rd and even 4th look, I've started this more than a year ago with long pauses and that's the perfect recipe for bugs...

ev-br commented:
Don't wait for me on this one --- I'm not really familiar with the subject area, so I won't be able to give it a meaningful review.

nmayorov commented:
Some thoughts on this PR:

1. TeX formulas look great in rendered html, but are very hard to read in text format. I believe it's better to avoid LaTeX in the main part of the docstring (and as far as I know it is a recommended practice in numpy/scipy).
2. `observer_hessenberg` counterpart is clearly missing.
3. More conceptual problem is what we are trying to achieve in `scipy.signal`. Say we add those basic functions for control system design and what next? We can't construct any controller or observer because `scipy.signal` doesn't provide any facilities for that. 
4. If we want to develop a useful package for control system design, than it should be probably separated to `scipy.control` (or `scipy.sytem`, or `scipy.lti_sys`) and provide more stuff than just checking observability or controllability. Although there is a `python-control` package available, but why not to integrate it (with some improvements and extensions) to scipy.

Generally I think that `scipy.signal` needs to be sorted out first, it is already quite convoluted and inconsistent. My feeling is that `scipy.signal` is more like `scipy.filter`, and everything related to LTI system design and analysis can be move to `scipy.system` (`scipy.lti_sys`).

I--P commented:
Hi,
1/ It's up to scipy devs to decide, once you'll agree on this point I'll change the docstring as needed
2/ I know, and it's wouldn't take a long time to add this function and it would provide us with a full Kalman decomposition but I have absolutely no time to do this right now, sorry.
3 and 4/ That's what I've been thinking for a long time... Otoh there is http://python-control.sourceforge.net/manual/intro.html which works great so I'm afraid there won't be many people willing to manage a scipy.control module and reimplement all the boring stuff already done in python-control.

rgommers commented:
> TeX formulas look great in rendered html, but are very hard to read in text format. I believe it's better to avoid LaTeX in the main part of the docstring (and as far as I know it is a recommended practice in numpy/scipy).

Correct. Sometimes it's needed, but in many cases it's better to stick to plain text in the main text, and keep TeX for the ``Notes`` section.

> It's up to scipy devs to decide .....

we *still* haven't made it official, but I'm definitely counting @nmayorov as belonging to that group:)

rgommers commented:
The scope of ``signal`` is definitely a concern. We know it's not quite right now, and we can either keep going adding functionality in the order in which people implement it, or we can say enough is enough and make a blueprint of how we want it to look a year from now and what's in/out of scope. The latter sounds tempting if we can find the time/energy for it.

nmayorov commented:
By straightforward considerations controllability and alike are out of scope of *signal processing* module. 

By all means I don't have anything against this code or its author (`controller_hessenberg` is a good function), but, in my opinion, `signal` needs clean-up and redesign first and this can go to a system design module (if it to appear).

All this could be a good project for GSoC, but unfortunately only next year now.

I--P commented:
_By all means I don't have anything against this code or its author (controller_hessenberg is a good function), In my opinion, signal needs clean-up and redesign first and this can go to a system design module (if it to appear)._

No worries I'm not taking this personally, actually I agree with you :-)

My only problem is that I'd like to have this PR closed and the code merged somewhere or discarded. I barely remember the algorithm (this was done more than a year ago) and if I have to wait another year to open another PR and merge it I'll for sure have forgotten everything.

rgommers commented:
Okay, so it looks like we won't be merging this anytime soon then. Removing the 0.18.0 milestone.

I had another good look at ``python-control``, and there is indeed a bit too much overlap between the LTI stuff in ``signal`` with that package. A longer-term plan would be good. About this PR, the functions here are useful and from what I can tell they are in better shape than what is in ``python-control``. So it would be a waste to throw this code away. Maybe it can be added to ``python-control`` instead. @murrayrm @cwrowley would this make sense?

murrayrm commented:
This would definitely fit within the scope of `python-control` and we would be happy to incorporate it into the code base if it can be rewritten as an appropriate PR.

ilayn commented:
I don't want to spark another discussion about this particular PR though I share the same confusion @I--P has in terms of contribution possibilities to `scipy.signal` in terms of its scope. Sometimes I see the bits and pieces of LTI system manipulations and I do see some opportunities to contribute in the `signal`. However, it is a bit unclear how to view `scipy.signal` module. 

Obviously, it is a difficult task to define the boundaries of such an interdisciplinary subject but at least in terms of intentions I believe we can cover some distance. For example, the scope and the intentions can be set from a working DSP engineer's toolbelt or it can be set as *anything needed to design IIR/FIR filters and signal manipulations* and so on (please take these propositions as raw drafts they don't imply anything). 

Because indeed `python-control` and some other libraries have similar initiatives but if `scipy.signal` has ambitions to become a full-fledged module in LTI systems then it can be an idea to focus the efforts on SciPy. 

I know that this is not ideal to discuss under this item which has a specific purpose, hence I would like to ask a concrete question: 

Is it possible to formulate an official long-term plan and where it is headed towards? This would help us the control people to make a better decisions of what should be a valid contribution for SciPy `signal` and what is too-specialized and be kept for 3rd party modules. 

Currently, I'm extrapolating only by the [Wiki to-do list](https://github.com/scipy/scipy/wiki/scipy.signal-to-do-list)

-----------

Regarding the code itself of this PR, it requires a more robust check of the staircase blocks as it is numerically not robust for weakly-controllable systems. You can find Boley's cancellation distance metric in the same book to judge whether the staircase algorithm is succesfully exited. Also the `Bunch()` object type is a bit unconventional. If I can, I would argue for this PR getting closed related to the roadmap discussion above as it would require a complete overhaul.

@I--P instead if you are still interested, I would really appreciate your feedback/contribution on the [implementation I have here](https://github.com/ilayn/harold/blob/master/harold/_system_funcs.py#L289)

I--P commented:
TBH I don't have much time to do scipy stuff anymore, historically I made place_poles to avoid installing python-control on 200+ computers, this PR was only to add a check for uncontrollable matrix I was missing in place_poles and it took me way too much time for no result at all. I'm in no way specialized in numerical computation so I'll let other people review your code but I would really like to see the staircase algorithm finally available in scipy.

The Bunch() object was an idea of @rgommers and I like it too, I avoids breaking everybody's code when you want to add  something in the returned values and it's more readable to write for example my_system.gain_matrix rather than my_system[2].

ilayn commented:
@I--P Don't get me wrong I do enjoy using pole placement code 👍 

Perhaps I should have added as : "If this is out of scope then let's define the scope a bit better and close this otherwise I can take it over to revamp"

I--P commented:
And don't get me wrong too :-) 
I would be very glad to see this PR revamped and incorporated in scipy and as I have zero time to do it it's a very good news that someone else is interested. Clearly this feature is missing in place_poles (and maybe some other functions ?) and I never understood why place_poles was "in scope" but a check on its arguments to make it more robust was out of scope.

Eric89GXL commented:
I'd be in favor of creating a new `scipy.control` (or `scipy.system`) module, and copying things there and deprecating them in `scipy.signal`. This sounds like what you proposed right @nmayorov? @rgommers does that make sense to you? There would be some overlap with `scipy.signal` / `scipy.filter` (if we refactor it again) but it seems manageable. If we can agree that's a reasonable approach, then this PR can be redone on top of an initial refactoring PR, hopefully without too much additional work.

murrayrm commented:
There has been some discussion among the `python-control` developers, including @cwrowley, about incorporating `python-control` into `scipy`.  If a `scipy.control` module was created, it should be possible to move move code from `python-control` into `scipy.control` over time.  The main issue from the `python-control` developers perspective is how much time it would take to maintain a `scipy.control` module.

Juanlu001 commented:
I used to think that `python-control` had licensing issues but apparently I was wrong, just the `slycot` project is GPLv2. And btw I just discovered @ilayn 's harold, nice job!

It would be great to have a `scipy.control` package. The next step would then be splitting the two modules IIUC, how could this be done in a scipy-compliant backwards compatible way?

Eric89GXL commented:
Well before we think too hard about the deprecation issues, let's see if other people are on board. It's probably worth an email to the scipy-dev mailing list to get discussion and feedback going, preferably by whomever would be interested in doing the heavy lifting on creating and maintaining the new submodule.

e-q commented:
I'm a user and a big fan of `python-control`, but I don't think I'm quite on board with merging it into scipy. The scope of capabilities that users may expect from a controls package is a little bigger than what I imagine for scipy submodules. I think there is an advantage to be had to being a standalone module that can set its own schedule, deprecation policy, etc.

cclauss commented:
Do we still need this PR?  If so, please rebase to fix the conflicts.

murrayrm commented:
If this PR doesn't get merged into `scipy.signal`, I'd like to see the changes captured in `python-control` (especially `controller_hessenberg`, which represents new functionality).

ilayn commented:
One thing that we first need to include for Hessenberg form is to use `dgehrd` instead of the Python native code. Also if we really want anybody use these matrices numerically we need a arnoldi step to normalize in order to avoid blowups,

I--P commented:
Sorry but I started this PR four (!!!) years ago and I don't have time, nor willingness, to rebase/fix it anymore, honestly I thought it had been ditched years ago. I can't even remember how the underlying algorithm works.

cclauss commented:
Then please close it?

ev-br commented:
Tagging it with needs-champion, so that if someone is willing to finish it up, it's up for grabs.

Juanlu001 commented:
Does it need a champion though? Older (2016) comments in this thread suggest that this is out of scope for `scipy.signal`, and that it should be split into `scipy.filter` and `scipy.control`. From my point of view, this needs a decision from SciPy developers first.

ilayn commented:
There is this positive generic feeling that `scipy.signal` should become more production ready in terms of LTI components. I think it is safe to say that we all feel that way. It is indeed showing its naivity in terms of underlying code and feature set.

But the counter argument that if we are to separate and create these new submodules, they can already become scikits right away which is kind of the actual situation with various packages python-control, harold, pyfda etc. 

Hence from the maintenance point of view, status quo seems the way to me (emphasis on *me* so not the official response). But then where do we draw to feature line and why there are then the immediate questions.

Juanlu001 commented:
Totally agree @ilayn. That's why "needs-decision" seems more appropriate now than "needs-champion" (and perhaps a thread in the mailing list, plus a review of https://github.com/scipy/scipy/blob/master/doc/source/roadmap-detailed.rst#signal ?).

larsoner commented:
Agreed, changed to `needs-decision`

rgommers commented:
Follow-up to this discussion about python-control and the scope/state of `scipy.signal` here: https://mail.python.org/pipermail/scipy-dev/2021-February/024521.html
